main/src/library/MutList.flix

```diff
@@ -354,6 +370,8 @@ namespace MutList {
     ///
     /// The result is a new mutable list.
     ///
+    /// pub def map(f: a -> b \ { ef - Write(r) }, v: MutList[a, r]): MutList[b, r] \ { ef, Read(r), Write(r) }
```

Comment:
Map is more interesting. 

Here the (arbitrary?) design choices seems to have been that `map` allocates a new `MutList` but in the *same* region. 

Question 0: Argue for/against this design choice.

Question 1: What would `map` look like if it took a region parameter and allocated them new `MutList` in that region? Write the signature!

Question 2: Does your answer change depending on whether `r` may alias `r2`. How would you deal with this?

Question 3: Since `map` actually allocates a new array (l366) does that change your answer whether `f` can have an effect in `r`/`r2`?

Reply:
0. Making a new array of the new type makes sense for type-checking.  
It is sensible unless we _really_ care about it being in-place, which I feel would require some shenanigans to type correctly and not create issue for previous references to the underlying array.

1. 
    ```scala 
    pub def map(f: a -> b \ { ef - Write(r1) }, v: MutList[a, r1], r2: Region[r2]): MutList[b, r2] \ { { ef - Write(r1) } , Read(r1), Write(r2) }
    ```

2. At the face of it, I think above suggestion would work, even with aliasing.

3. `f` having an effect in `r2` should be fine, as I do not believe it can reference the memory we are working on in `r2`.

__Counter Question:__  
Since we are making a new array anyway. Why not make use the functional implementation on lists, whereby `f` can do whatever it wants without issue?

```scala
pub def map(f: a -> b \ ef, v: MutList[a, r]): MutList[b, r] \ { ef, Read(r), Write(r) } =
        let r = Scoped.regionOf(v);
        if (isEmpty(v))
            new MutList(r)
        else {
            let MutList(ra, rl) = v;
            let a = toList(deref ra);
            let l = deref rl;
            let b = List.map(f, a);
            List.toMutList(r, b)
        }
```